Wanhao duplicator i3

mtpiercey opened the pull request:
### Description

While the Wanhao Duplicator i3 Mini is represented in the example configurations, the v2.1 (which is quite different, actually) is not. As suggested by [@zerog2k](https://github.com/BotInABoxER/marlin2-for-wanhao-i3/issues/3#issue-490685831), I prepared my port to be added to the list.

I added an example `Configuration.h`, `Configuration_adv.h`, and a `_Bootscreen.h` specific to the i3 2.1. I also added a thermistor table, since none of the presets seemed to work (though I don't think any of the other examples have their own tables), and only after generating one specific to my board's thermistor did I get the results I wanted.

I also added a `README.md` in the folder with a bit of explanation, and a `Chippy_Bootscreen.h` from Bot-In-a-Box Educational Robotics, because I think it looks pretty slick.

### Benefits

This adds support for the Wanhao Duplicator i3 v2.1 via creating an example configuration file (something that I would have appreciated months ago).

mtpiercey committed:
Add config for Wanhao Duplicator i3 v2.1

mtpiercey committed:
Add newlines to Wanhao i3 v2.1 config files

mtpiercey committed:
Add Chippy Bootscreen and tweak README for Wanhao i3 2.1 config

mtpiercey committed:
Tweak README for Wanhao i3 v2.1 config

zerog2k commented:
Should thermistor table get a dedicated id and merged into the rest of the tables? (Do maintainers have a suggestion on this, and if so, an ID which can be used, i.e. next available, monotonically increasing table id, e.g. `14`)
edit: I'm suggesting this thermistor table be added along side rest of them rather than some separate, custom thing, if that makes sense?

thinkyhead committed:
Update Chippy_Bootscreen.h

thinkyhead committed:
Update _Bootscreen.h

thinkyhead commented:
If the thermistor is truly unique then it can get its own ID. But it is definitely preferred to stick with the closest thermistor that applies, and then add instructions in the README on how to calibrate your own thermistor. If the Wanhao thermistors are somehow known too all be similar to each other, then it's not a great burden to add one more oddball thermistor to the pile. If possible, we should get confirmation on whether the measured thermistor is a good match for other Wanhao i3 machines.

zerog2k commented:
I have Wanhao Duplicator i3 V2.1 (Melzi "v4.4" board), and confirm this works ok. This board definately has the strange 100K thermistor with 10K resistor (on top of divider?)
However, I'm also going through the process of calibrating extruder temps of my own machine (roughly following https://youtu.be/brl1weuD8bs), and finding that, per my Fluke 16 thermocouple [TC] (which was sanity checked to boiling point of water ;) read about 20 C lower than what Marlin things temp is (per the thermistor table provided here).

@mtpiercey can you share your methodology for obtaining your thermistor table?

However I am not sure if I have the best thermal coupling of the TC to the extruder - due to the size of my TC bead, I'm unable to stick it inside the nozzle itself, but rather "squeeze" the TC onto a glass bed. I'm suspecting due to ambient airflow and thermal "sinking"(?) of glass/bed, this could be skewing results slightly. I'm thinking further of how I can isolate some of these source of errors, like maybe pressing into a silicone pad or something.

As an aside, this highlights that using 10k (instead of the more common 4.7k) is a really strange choice by designers, as around 200C+ the resistance on a 100k thermistor is about 500ohms, so the 10k resistor pushes the "interesting" region of our measurements into the much steeper parts of the curve. At some point, I may be tempted to mod the board and replace that resistor to give better resolution around the desired temp band.

mtpiercey commented:
So, I've heard that the 4.7k seems to be an upgrade of choice for those for this particular printer, and apparently some of the "newer" v2.1s came with them from the factory (though I can't confirm that).

It does seem to be that the 10k thermistors are at least prevalent, but I'm not sure to which degree.

Like I mentioned in the README, I generated the table with `buildroot/share/scripts/createTemperatureLookupMarlin.py --rp=10000`, and it just happened to work pretty well. I can confirm that, as far as I have been able to measure, the table makes the readings accurate with a tolerance of about one Centigrade degree.

I'd personally suggest adding a custom table only if the actual measurements are done to verify/tweak it, though, since I'm no engineer. However, I can verify the off-by-20° C phenomenon, since before I generated the 10k table I was getting readings of ~40° C at room temperature (with was around 21° C at the time).

zerog2k commented:
> table makes the readings accurate with a tolerance of about one Centigrade degree

I'm curious how you measured this (e.g. thermocouple inside nozzle, taped to block, IR gun, etc) ?

Anyhow, I would suggest that this is worthy of a new thermistor table. 
There is another 100k thermistor, 10k pull-up resistor table, `thermistor_8.h`,  however looking at it, the calibration points look more for a hot bed (there arent enough sample points in the table around normal extrusion temps, i.e. 180-240) to provide smaller linear fit segments to the (very) sharp curve at that point. Hence, I agree it deserves it's own table. 
I expect individuals (like myself) who have the tools to measure and further calibrate it locally can do so to our liking.

(Thanks for this @mtpiercey !)

thinkyhead commented on config/examples/Wanhao/Duplicator i3 2.1/thermistors.h:
```diff
@@ -0,0 +1,325 @@
+#endif
+#if ANY_THERMISTOR_IS(99) // 100k bed thermistor with a 10K pull-up resistor (on some Wanhao i3 models)
+  #include "thermistor_99.h"
+#endif
```
You can just put this entry into the main `thermistors.h` and delete this file.

thinkyhead replied:
You'll also need to update `thermistornames.h` and add thermistor 99 to all the `Configuration.h` files.

mtpiercey commented:
Thanks @thinkyhead for the suggestion, I'll look into editing those files tonight or tomorrow. Also, @zerog2k, I just used a simple digital thermometer to measure the temperature in another room and compared it to what the printer said. Very low-tech and imprecise in that regard.

mtpiercey committed:
Add thermistor 99 for some Wanhao i3 models

mtpiercey committed:
Add thermistor 99 to base Configuration.h

mtpiercey commented:
OK, I think I did that right.

thinkyhead committed:
Update configs for thermistor 99

...and remove the unmaintained JSON array. (Configurators can parse the comment instead.)

thinkyhead commented:
I've submitted a PR to finish this up. Please merge BotInABoxER/Marlin#1 when you have a chance.

mtpiercey committed:
Merge pull request #1 from thinkyhead/bf2_wanhao_i3_15330

Update configs for thermistor 99…

thinkyhead commented:
This has been merged, but I had to do it directly because there were merge conflicts and this PR was created without giving maintainers permission to edit it.

mtpiercey commented:
OK, cool. Guess I should have given those write permissions off the bat (thanks for the tip).

Looks like we're done here.
